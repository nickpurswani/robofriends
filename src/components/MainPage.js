import React, { useEffect,useState } from "react";
import CardList from "./CardList";
import SearchBox from "./SearchBox";
import Scroll from "./Scroll";
import Header from './Header';
import ErrorBoundary from "./ErrorBoundary";

const MainPage=(props)=>{
    const [loading,setLoading]=useState(true);
    const {searchField,onSearchChange,robots,onRequestRobots}=props;
    useEffect( ()=>{
        const fetchData = async () => {
            try {
              await onRequestRobots();
              
            } catch (error) {
              // Handle errors if necessary
              console.error('Error fetching robots:', error);
            } finally {
              setLoading(false);
            }
          };
        
          fetchData(); 
    },[onRequestRobots])
        const filteredRobots=robots.filter(robots =>{
            return robots.name.toLowerCase().includes(searchField.toLowerCase());    
            });
           
            if (loading){
                return <h1>.....Loading</h1>
            }
            else{
    return(
        <div className="tc">
            <Header/>
            <SearchBox searchchange={onSearchChange}/>
            <Scroll>
            <ErrorBoundary>
            <CardList robots={filteredRobots}  />
            </ErrorBoundary>
            </Scroll>
        </div>
    )}

}
export default MainPage