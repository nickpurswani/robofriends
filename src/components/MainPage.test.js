import React from 'react';
import { render, screen, waitFor,act  } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MainPage from './MainPage';
import '@testing-library/jest-dom'


const mockProps = {
    onRequestRobots: jest.fn(() => Promise.resolve([
      {
        id: 1,
        name: 'Leanne Graham',
        username: 'Bret',
        email: 'Sincere@april.biz',
      },
      {
        id: 2,
        name: 'John Doe',
        username: 'john_doe',
        email: 'john.doe@example.com',
      },

     
    ])),
    robots: [{
        id: 1,
        name: 'Leanne Graham',
        username: 'Bret',
        email: 'Sincere@april.biz',
      },
      {
        id: 2,
        name: 'John Doe',
        username: 'john_doe',
        email: 'john.doe@example.com',
      }], // Initial empty state
    searchField: '',
    onSearchChange: jest.fn(), // Mock onSearchChange function
  };
  test('renders MainPage with loading message while fetching data', async () => {
    const { rerender } = render(<MainPage {...mockProps}  />);
    
    expect(screen.getByText('.....Loading')).toBeInTheDocument();
    let fetcheddata;
    await act(async () => {
      // Simulate the asynchronous operation
     await mockProps.onRequestRobots();
    });
    rerender(<MainPage {...mockProps} loading={false} />);
    expect(screen.queryByText('.....Loading')).not.toBeInTheDocument();
    expect(screen.queryByText('.....Loading')).not.toBeInTheDocument();
    expect(screen.getByText('Leanne Graham')).toBeInTheDocument();
    expect(screen.getByText('John Doe')).toBeInTheDocument();
  });
test('filters robots with  search field', async () => {
    const { rerender } = render(<MainPage {...mockProps} />);
    
    expect(screen.getByText('.....Loading')).toBeInTheDocument();
    let fetcheddata;
    await act(async () => {
      // Simulate the asynchronous operation
     await mockProps.onRequestRobots();
    });
    rerender(<MainPage {...mockProps}searchField="Leanne" loading={false} />);
    expect(screen.queryByText('.....Loading')).not.toBeInTheDocument();
    expect(screen.queryByText('.....Loading')).not.toBeInTheDocument();
    expect(screen.getByText('Leanne Graham')).toBeInTheDocument();
    expect(screen.queryByText('John Doe')).not.toBeInTheDocument();
  });