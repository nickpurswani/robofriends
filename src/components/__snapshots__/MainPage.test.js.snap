// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="tc"
  >
    <div>
      <h1
        class="f1"
      >
        RoboFriends
      </h1>
    </div>
    <div>
      <input
        class="pa3 ba b--green bg-lightest-blue"
        id="seachbox"
        placeholder="search robots"
        type="search"
      />
    </div>
    <div
      style="overflow: scroll; border: 5px solid black; height: 800px;"
    >
      <h1>
        Error occured while fetching robofriends
      </h1>
    </div>
  </div>
</DocumentFragment>
`;
